### PR TITLE
fix Stm32 UUID

### PR DESCRIPTION
**Pull request by ellensp**
### Description

The OPULO_LUMEN_REV3 and OPULO_LUMEN_REV4 boards use the inbuilt STM32 
CPU serial number as a non standard default UUID  (its only 24 digits not 32 and doesn't use standard 8-4-4-4-12 format)  

At some stage DEFAULT_MACHINE_UUID became required, breaking this feature.
I have presumed that there is some reason to have a mandatory DEFAULT_MACHINE_UUID and worked around it.

### NOTE

This works, but i not 100% happy with it.

Ideally need to find a way to set NEED_HEX_PRINT when HAS_STM32_UID is set in the pins file, as this features uses HEX_PRINT 
Currently I have added MB(OPULO_LUMEN_REV3) || MB(OPULO_LUMEN_REV4) to the test that sets NEED_HEX_PRINT
Otherwise HEX_PRINT code is filtered out and it will not build.

### Requirements

HAS_STM32_UID In pins file 
MB(OPULO_LUMEN_REV3) || MB(OPULO_LUMEN_REV4)

### Benefits

Works as expected

### Configurations

From he issue https://github.com/MarlinFirmware/Marlin/files/13966641/Archive.zip

### Related Issues

https://github.com/MarlinFirmware/Marlin/issues/26698

### CODE REVIEW

**Comment by ellensp** [comment]
As suggested in the issue comments moving HAS_STM32_UID to the Configuration.h would allow automatic setting of NEED_HEX_PRINT based on the ENABLED(HAS_STM32_UID)

**Comment by thisiskeithb** [comment]
> As suggested in the issue comments moving HAS_STM32_UID to the Configuration.h would allow automatic setting of NEED_HEX_PRINT based on the ENABLED(HAS_STM32_UID)

I think it'd be handy to set the UUID to the board SN, especially since most (all?) STM32-based boards support it.

**Comment by ellensp** [comment]
Just noticed that this UUID is also in uppercase, most seem to be written in lowercase.  Google says case doesnt matter. 
I also think we should format it to 8-4-4-4-12 format with a prefix of "eebedded"  close as can get to embedded

ie if serial number is deadbeafdeadbeafdeadbeaf  we should display
eebedded-dead-beaf-dead-beafdeadbeaf

**Comment by thinkyhead** [comment]
> As suggested in the issue comments moving `HAS_STM32_UID` to the `Configuration.h`

Better not to have conditional names (i.e., `HAS_*` etc.) set by user configuration, and this conditional makes more sense as it is, pertaining to the MCU model or board.

So I simply moved `NEEDS_HEX_PRINT` to `Conditionals_post.h`.
- If the user supplies a `MACHINE_UUID` it overrides everything else,
- otherwise, if `HAS_STM32_UID` it falls back to that,
- and if `!HAS_STM32_UID` it falls back to the `DEFAULT_MACHINE_UUID` as defined by `Version.h`.

> I think it'd be handy to set the UUID to the board SN

Sure, I have no problem with making `HAS_STM32_UID` be set for all STM32 boards if that works.

**Comment by thinkyhead** [comment]
> Just noticed that this UUID is also in uppercase, most seem to be written in lowercase. 

In this case we have to use whatever the Opulo Lumen looks for. So what is that?

**Comment by thinkyhead** [comment]
> Google says case doesnt matter

I agree. It could be just about anything and that would be fine as long as it serves the needs of your workshop.

**Comment by ellensp** [comment]
This is not working as intended...

if you temporarily change 

#if ANY(AUTO_BED_LEVELING_UBL, M100_FREE_MEMORY_WATCHER, DEBUG_GCODE_PARSER, TMC_DEBUG, MARLIN_DEV_MODE, DEBUG_CARDREADER, M20_TIMESTAMP_SUPPORT, HAS_STM32_UID)

to 

#if ENABLED(HAS_STM32_UID)

and remove #define TMC_DEBUG from Configuration_adv.h

Your back to
Marlin/src/gcode/host/M115.cpp:92: undefined reference to `hex_word(unsigned short)'

the only reason the examples/Opulo builds is it sets NEED_HEX_PRINT as it  has #define TMC_DEBUG in the Configuration_adv.h.  HAS_STM32_UID is not triggering this

File: Marlin/src/gcode/host/M115.cpp [hunk 1]
```diff
@@ -71,24 +72,31 @@ void GcodeSuite::M115() {
...
-    }
+  #if !defined(MACHINE_UUID) && HAS_STM32_UID
+    /**
+     * STM32-based devices have a 96-bit CPU device serial number.
+     * Used by LumenPnP / OpenPNP to keep track of unique hardware/configurations.
+     * https://github.com/opulo-inc/lumenpnp
+     * This code should work on all STM32-based boards.
+     */
+    #if DISABLED(STM32_UID_GNF)
+      uint16_t * const UID = (uint16_t*)UID_BASE;
```
Comment (classicrocker883) [inline Marlin/src/gcode/host/M115.cpp]: `STM32_UID_GNF` - is that a typo?

File: Marlin/src/gcode/host/M115.cpp [hunk 2]
```diff
@@ -71,24 +72,31 @@ void GcodeSuite::M115() {
...
+      SERIAL_ECHO(
+        F("CEDE2A2F-"), hex_word(UID[0]), '-', hex_word(UID[1]), '-', hex_word(UID[2]), '-',
+        hex_word(UID[3]), hex_word(UID[4]), hex_word(UID[5])
```
Comment (classicrocker883) [inline Marlin/src/gcode/host/M115.cpp:97]: `hex_word` undefined here

File: Marlin/src/gcode/host/M115.cpp (hunk 1 above)
Comment (ellensp) [reply Marlin/src/gcode/host/M115.cpp]: I think it should be HAS_STM32_UID

File: Marlin/src/gcode/host/M115.cpp (hunk 2 above)
Comment (ellensp) [reply Marlin/src/gcode/host/M115.cpp:97]: this is just a symptom of the broken source filtering, so the code for  hex_word or hex_long is not compiled in.

**Comment by ellensp** [comment]
oops, test configs where in error.

so what is this STM32_UID_GNF then?

**Comment by thinkyhead** [comment]
> so what is this STM32_UID_GNF then?

Set this if you give no f***s.

File: Marlin/src/gcode/host/M115.cpp (hunk 2 above)
Comment (thinkyhead) [reply Marlin/src/gcode/host/M115.cpp:97]: Nope, it's a symptom of strange type identification. See the commit that solves it, just added.

**Comment by thisiskeithb** [comment]
> > so what is this STM32_UID_GNF then?
> 
> Set this if you give no f***s.

Could we use a name that is a bit more...descriptive?

**Comment by thinkyhead** [comment]
> Could we use a name that is a bit more...descriptive?

No worries. I'll rename it when it stops being amusing.

**Pull request merged**